nixos/modules/services/matrix/synapse.nix

```diff
@@ -154,7 +197,108 @@ in {
 
   ];
 
-  options = {
+  options = let
+    listenerType = types.submodule {
+      options = {
+        port = mkOption {
+          type = types.port;
+          example = 8448;
+          description = lib.mdDoc ''
+            The port to listen for HTTP(S) requests on.
+          '';
+        };
+
+        bind_addresses = mkOption {
+          type = types.listOf types.str;
+          default = [
+            "::1"
+            "127.0.0.1"
+          ];
+          example = literalExpression ''
+            [
+              "::"
+              "0.0.0.0"
+            ]
+          '';
+          description = lib.mdDoc ''
+            IP addresses to bind the listener to.
+          '';
+        };
+
+        type = mkOption {
+          type = types.enum [
+            "http"
+            "manhole"
+            "metrics"
+            "replication"
+          ];
+          default = "http";
+          example = "metrics";
+          description = lib.mdDoc ''
+            The type of the listener, usually http.
+          '';
+        };
+
+        tls = mkOption {
+          type = types.bool;
+          default = true;
```

Comment:
There's no particular reason for this default as far as I can tell. I didn't change anything here, I just extracted the type definition for listeners to be able to reuse it for `services.matrix-synapse.settings.listeners` and `services.matrix-synapse.workers.*.worker_listeners`.

If workers are enabled, you have to terminate TLS at a reverse proxy anyway, so it'd probably make sense to default it to `false` anyway including for the main process. What do you think?

Reply:
> If workers are enabled, you have to terminate TLS at a reverse proxy anyway, so it'd probably make sense to default it to false anyway including for the main process. What do you think?

Correct. However, I think another common way to deploy is to actually expose port 8448 publicly (for federation in non-worker context) which needs to speak TLS then. Anyways, I don't really want to change the default for the main process's listener, I'm afraid that this might be too easy to miss on upgrade.

I agree however, there's no reason to do TLS for workers, so this default needs to change. I.e.

```nix
let
  listenerType = workerContext: types.submodule {
    options.tls = mkOption {
      default = !workerContext;
      /* ... */
    };
  };
in
```

and then `listenerType true` for `worker_listeners.type` and `listenerType false` for `listeners.type`.

Reply:
Done.